File: tempestweb/native/imaging.py
"""Compress, thumbnail and transform an image before it is uploaded (R6).

Between :func:`tempestweb.native.camera.capture` and
:func:`tempestweb.native.http.upload` there was nothing. An app captured a 4 MB
photo and uploaded 4 MB, or rewrote canvas compression by hand — in a framework
whose proposal is not writing JS.

**The pixels stay in the browser.** Every function here takes and returns an
:class:`ImageRef`, an opaque handle to bytes the client is holding. Python
addresses the image by name; the image never crosses the bridge:

```text
Mode B, a 4 MB photo, compressing it:

  bytes:   client →5.3MB→ server →5.3MB→ client   (10.6 MB of network)
  handle:  client →"blob:tw:7"→ server →"blob:tw:7"→ client   (~40 bytes)
```

Example:
    ```python
    from tempestweb import native


    async def upload_photo() -> None:
        photo = await native.camera.capture(include_bytes=False)
        small = await native.imaging.compress(photo, max_kb=200, max_width=1600)
        print(small.size_kb, small.quality, small.attempts, small.within_budget)
        await native.http.upload("/api/fotos", small.as_upload("foto.jpg"))
    ```

The part nobody gets right on the first try is the compression: it is a **binary
search of encoder quality against a byte budget**, because encoded size is not
linear in quality. :class:`CompressedImage` reports where it landed — which
quality, how many encodes it spent, and whether it met the budget at all.

!!! warning "An impossible budget answers, it does not hang"
    Asking for 200 KB of a photo that will not go below 400 KB at the floor
    quality gives ``within_budget=False`` and the smallest it managed, after a
    **bounded** number of encodes. A too-large image the app can decide about
    beats a spinner that never stops.

!!! note "Handles are bounded, and an evicted one is a named error"
    The client holds at most a few dozen blobs and evicts the oldest, so a photo
    app running for an hour does not accumulate every frame. Addressing an
    evicted handle raises ``NativeError("not_found")`` — recover by capturing
    again, not by retrying.
"""

from __future__ import annotations

import base64
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from tempestweb.native.dispatch import send_native_call

__all__ = [
    "ImageRef",
    "ImageSource",
    "CompressOptions",
    "TransformOptions",
    "CropBox",
    "CompressedImage",
    "Thumbnail",
    "ProcessedImage",
    "ImageInfo",
    "ImageBytes",
    "AUTO_TYPE",
    "compress",
    "thumbnails",
    "transform",
    "info",
    "read",
    "release",
]

#: Ask the browser to pick the smallest type it can encode (WebP where
#: available, JPEG otherwise). One field instead of the two capabilities the
#: React SDK spends on it: the caller wants a small file, not a survey.
AUTO_TYPE = "auto"

#: An opaque handle to bytes the client is holding. Never parsed here — its shape
#: is the client's business, and Python only ever hands it back.
ImageRef = str

#: What a capability accepts as an image: a handle, or any model carrying one
#: (a :class:`~tempestweb.native.camera.Photo`, a
#: :class:`~tempestweb.native.file.PickedFile`, or a result from this module).
ImageSource = "ImageRef | BaseModel | dict[str, Any]"


class _Payload(BaseModel):
    """Base for a model the **browser** fills in.

    Extras are ignored on purpose: a newer client sending a field an older Python
    does not know about must not break that Python. The opposite rule applies to
    the option models below.
    """

    model_config = ConfigDict(frozen=True)


class _Options(BaseModel):
    """Base for a model the **developer** fills in.

    ``extra="forbid"`` because a misspelled option is a bug and has to hurt. The
    alternative — silently ignoring ``max_width`` because it was typed
    ``maxWidth`` — uploads a full-size photo and says nothing.
    """

    model_config = ConfigDict(frozen=True, extra="forbid")


class CropBox(_Options):
    """A rectangle to cut out of the source, in source pixels.

    Attributes:
        x: Left edge.
        y: Top edge.
        width: Rectangle width.
        height: Rectangle height.
    """

    x: int = 0
    y: int = 0
    width: int = 0
    height: int = 0


class CompressOptions(_Options):
    """How hard to compress, and what to aim for.

    Attributes:
        max_kb: The byte budget in kilobytes. ``None`` compresses without a
            target, which just re-encodes at ``max_quality``.
        max_width: Cap the width, keeping the aspect ratio. Never upscales.
        max_height: Cap the height, keeping the aspect ratio. Never upscales.
        type: The output MIME type, or :data:`AUTO_TYPE`.
        min_quality: The floor the search will not go below.
        max_quality: The ceiling the search starts from.
        steps: How many encodes the search may spend. Bounding this is what makes
            an impossible budget answer instead of hang.
    """

    max_kb: float | None = None
    max_width: int | None = None
    max_height: int | None = None
    type: str = AUTO_TYPE
    min_quality: float = 0.4
    max_quality: float = 0.92
    steps: int = 6


class TransformOptions(_Options):
    """Resize, rotate, crop and flip, in one pass over the image.

    One call rather than four, because four means decoding and encoding the same
    image four times to do what one canvas does once.

    Attributes:
        width: Cap the width, keeping the aspect ratio.
        height: Cap the height, keeping the aspect ratio.
        rotate: Degrees clockwise. 90 and 270 swap the output's axes.
        flip_horizontal: Mirror left-to-right.
        flip_vertical: Mirror top-to-bottom.
        crop: A rectangle to cut out before the rest is applied.
        type: The output MIME type, or :data:`AUTO_TYPE`.
        quality: The encoder quality for the single encode.
    """

    width: int | None = None
    height: int | None = None
    rotate: int = 0
    flip_horizontal: bool = False
    flip_vertical: bool = False
    crop: CropBox | None = None
    type: str = AUTO_TYPE
    quality: float = 0.92


class _Handled(_Payload):
    """A payload carrying a handle to bytes the client is holding."""

    ref: ImageRef = ""
    mime_type: str = "application/octet-stream"

    def as_upload(self, name: str) -> dict[str, Any]:
        """Describe this image for :func:`tempestweb.native.http.upload`.

        The whole point of the handle: the upload reads the bytes on the client,
        so they are sent once, to the server that wanted them — not once to
        Python and again to the server.

        Args:
            name: The file name to send.

        Returns:
            The file descriptor ``http.upload`` expects.
        """
        return {"name": name, "type": self.mime_type, "blob_id": self.ref}


class CompressedImage(_Handled):
    """What the quality search settled on.

    Attributes:
        ref: The handle to the compressed bytes.
        mime_type: What it was encoded to — useful when ``type`` was ``"auto"``.
        width: The output width in pixels.
        height: The output height in pixels.
        size_kb: The output size in kilobytes.
        quality: The encoder quality the search chose.
        attempts: How many encodes it spent getting there.
        within_budget: Whether ``max_kb`` was met. **Check this** — ``False``
            means the budget was impossible and this is the smallest it managed.
    """

    width: int = 0
    height: int = 0
    size_kb: float = 0.0
    quality: float = 0.0
    attempts: int = 0
    within_budget: bool = False


class Thumbnail(_Handled):
    """One rendered preview.

    Attributes:
        ref: The handle to the thumbnail's bytes.
        mime_type: What it was encoded to.
        size: The requested maximum edge, echoed back so a caller can match a
            thumbnail to the size it asked for without relying on order.
        width: The actual width, after fitting the aspect ratio.
        height: The actual height.
        size_kb: The size in kilobytes.
    """

    size: int = 0
    width: int = 0
    height: int = 0
    size_kb: float = 0.0


class ProcessedImage(_Handled):
    """The result of a transform.

    Attributes:
        ref: The handle to the processed bytes.
        mime_type: What it was encoded to.
        width: The output width.
        height: The output height.
        size_kb: The output size in kilobytes.
    """

    width: int = 0
    height: int = 0
    size_kb: float = 0.0


class ImageInfo(_Payload):
    """What an image is, without moving it.

    Attributes:
        mime_type: The image's MIME type.
        width: The intrinsic width in pixels.
        height: The intrinsic height in pixels.
        size_kb: The byte size in kilobytes.
    """

    mime_type: str = "application/octet-stream"
    width: int = 0
    height: int = 0
    size_kb: float = 0.0


class ImageBytes(_Payload):
    """An image pulled back into Python.

    Attributes:
        data_base64: The bytes, base64-encoded.
        mime_type: The image's MIME type.
        size_kb: The byte size in kilobytes.
    """

    data_base64: str = Field(default="", repr=False)
    mime_type: str = "application/octet-stream"
    size_kb: float = 0.0

    def to_bytes(self) -> bytes:
        """Decode to raw bytes.

        Returns:
            The decoded image bytes.
        """
        return base64.b64decode(self.data_base64)


def _source(source: object) -> object:
    """Render whatever the caller passed as the wire's ``source`` field.

    Args:
        source: A handle, a model carrying one, or a payload carrying base64.

    Returns:
        The handle when one can be found, otherwise the value itself — the client
        also accepts ``{data_base64, mime_type}``, which is what
        ``camera.capture`` and ``file.pick`` answered before handles existed.
    """
    if isinstance(source, str):
        return source
    ref = getattr(source, "ref", None)
    if isinstance(ref, str) and ref:
        return ref
    if isinstance(source, BaseModel):
        return source.model_dump()
    return source


async def compress(
    source: object,
    *,
    options: CompressOptions | None = None,
    **overrides: Any,  # noqa: ANN401 — the CompressOptions fields, spelled inline
) -> CompressedImage:
    """Shrink an image to fit a byte budget, by searching encoder quality.

    Args:
        source: The image: a handle, a ``Photo``, a ``PickedFile``, or a previous
            result from this module.
        options: The options, built explicitly.
        **overrides: The same fields spelled inline —
            ``compress(photo, max_kb=200)``. A misspelled name **raises**, because
            silently ignoring ``maxWidth`` uploads a full-size photo and says
            nothing.

    Returns:
        The :class:`CompressedImage`. **Check ``within_budget``**: ``False`` means
        the budget could not be met and this is the smallest it managed.

    Raises:
        ValidationError: If an override is not a :class:`CompressOptions` field.
        NativeError: If the handle is unknown or evicted (``not_found``), the
            bytes are not an image (``decode_failed``), or the browser cannot
            encode the requested type (``encode_failed``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    resolved = options if options is not None else CompressOptions(**overrides)
    value = await send_native_call(
        "imaging.compress",
        {"source": _source(source), **resolved.model_dump(exclude_none=True)},
    )
    return CompressedImage.model_validate(value)


async def thumbnails(
    source: object,
    sizes: list[int],
    *,
    type: str = AUTO_TYPE,
    quality: float = 0.92,
) -> list[Thumbnail]:
    """Render one image at several sizes, for previews.

    Args:
        source: The image, in any form :func:`compress` accepts.
        sizes: The maximum edge of each thumbnail, in pixels.
        type: The output MIME type, or :data:`AUTO_TYPE`.
        quality: The encoder quality.

    Returns:
        One :class:`Thumbnail` per requested size, in the order asked. An empty
        ``sizes`` returns ``[]`` — asking for no thumbnails is valid, not an
        error.

    Raises:
        NativeError: If the handle is unknown or evicted (``not_found``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call(
        "imaging.thumbnails",
        {
            "source": _source(source),
            "sizes": list(sizes),
            "type": type,
            "quality": quality,
        },
    )
    raw = value.get("thumbnails", [])
    if not isinstance(raw, list):
        return []
    return [Thumbnail.model_validate(item) for item in raw]


async def transform(
    source: object,
    *,
    options: TransformOptions | None = None,
    **overrides: Any,  # noqa: ANN401 — the TransformOptions fields, spelled inline
) -> ProcessedImage:
    """Resize, rotate, crop and flip in a single pass.

    Args:
        source: The image, in any form :func:`compress` accepts.
        options: The options, built explicitly.
        **overrides: The same fields spelled inline. A misspelled name raises.

    Returns:
        The :class:`ProcessedImage`.

    Raises:
        ValidationError: If an override is not a :class:`TransformOptions` field.
        NativeError: If the handle is unknown or evicted (``not_found``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    resolved = options if options is not None else TransformOptions(**overrides)
    value = await send_native_call(
        "imaging.transform",
        {"source": _source(source), **resolved.model_dump(exclude_none=True)},
    )
    return ProcessedImage.model_validate(value)


async def info(source: object) -> ImageInfo:
    """Report an image's type, size and dimensions without moving it.

    Cheap by design: the byte count and MIME come from the blob itself. An app
    deciding whether an image is worth compressing at all asks this first.

    Args:
        source: The image, in any form :func:`compress` accepts.

    Returns:
        The :class:`ImageInfo`.

    Raises:
        NativeError: If the handle is unknown or evicted (``not_found``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    return ImageInfo.model_validate(
        await send_native_call("imaging.info", {"source": _source(source)})
    )


async def read(source: object) -> ImageBytes:
    """Pull an image's bytes back into Python.

    The escape hatch, **not** the path. This moves the whole image across the
    bridge, which is what handles exist to avoid — an app that only needs to
    upload should hand the handle to :func:`tempestweb.native.http.upload` via
    :meth:`CompressedImage.as_upload`.

    Args:
        source: The image, in any form :func:`compress` accepts.

    Returns:
        The :class:`ImageBytes`.

    Raises:
        NativeError: If the handle is unknown or evicted (``not_found``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    return ImageBytes.model_validate(
        await send_native_call("imaging.read", {"source": _source(source)})
    )


async def release(source: object | None = None, *, all: bool = False) -> int:
    """Release a handle, or every handle.

    Handles are evicted automatically once the registry is full, so this is an
    optimisation rather than a duty — worth calling when a screen holding several
    large images is left.

    Args:
        source: The handle to release.
        all: Release every handle instead.

    Returns:
        How many handles were released. Releasing one that is already gone
        answers ``0`` rather than raising — a double release is not an error.

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    args: dict[str, Any] = {"all": all}
    if source is not None:
        args["source"] = _source(source)
    value = await send_native_call("imaging.release", args)
    released = value.get("released", 0)
    return int(released) if isinstance(released, int | float) else 0

File: client/native/imaging.js
// native/imaging.js — compress, thumbnail and transform an image before upload.
//   PHASE R6.
//
// Between `camera.capture()` and `http.upload()` there was nothing: an app
// captured a 4 MB photo and uploaded 4 MB, or rewrote canvas compression by hand
// in a framework whose whole proposal is not writing JS.
//
// The part nobody gets right on the first try is `compressToTarget`: a BINARY
// SEARCH of encoder quality against a byte budget. Encoding is not linear in
// quality, so "try 0.8, then 0.6, then 0.4" either overshoots the budget or
// throws away quality that fit. The search converges in a bounded number of
// encodes and reports what it settled on — which quality, how many encodes it
// spent, and whether it met the budget at all.
//
// An impossible budget (200 KB of a photo that will not go below 400 KB at the
// floor quality) must NOT spin forever. The search is bounded by
// DEFAULT_COMPRESS_STEPS and answers `within_budget: false` with the smallest it
// managed, because a too-large image the app can decide about beats a hang.
//
// Everything here works on blob handles (native/blobs.js): the pixels never
// cross the bridge. Canvas is injectable via `deps.createCanvas` so the whole
// module is testable under jsdom, which has no canvas of its own.

import { CapabilityError } from "./index.js";
import {
  clearBlobs,
  describeBlob,
  dropBlob,
  putBlob,
  resolveSource,
  toBase64,
} from "./blobs.js";

/** How many encodes the quality search is allowed to spend. */
export const DEFAULT_COMPRESS_STEPS = 6;

/** The lowest quality the search will accept. Below this, artefacts show. */
export const DEFAULT_MIN_QUALITY = 0.4;

/** The highest quality the search starts from. */
export const DEFAULT_MAX_QUALITY = 0.92;

/** The type an `"auto"` request encodes to when the browser supports it. */
export const PREFERRED_TYPE = "image/webp";

/** The type every browser encodes. */
export const FALLBACK_TYPE = "image/jpeg";

/**
 * Resolve the canvas factory: injected for tests, `document` in a page.
 *
 * @param {import("./index.js").NativeDeps} deps
 * @returns {(w: number, h: number) => *} A factory returning a canvas-like object.
 * @throws {CapabilityError} unavailable — when there is no way to make a canvas.
 */
function canvasFactory(deps) {
  const injected = deps && /** @type {any} */ (deps).createCanvas;
  if (typeof injected === "function") return injected;
  const doc = (deps && deps.document) || /** @type {any} */ (globalThis).document;
  if (doc && typeof doc.createElement === "function") {
    return (w, h) => {
      const canvas = doc.createElement("canvas");
      canvas.width = w;
      canvas.height = h;
      return canvas;
    };
  }
  const g = /** @type {any} */ (globalThis);
  if (typeof g.OffscreenCanvas === "function") {
    return (w, h) => new g.OffscreenCanvas(w, h);
  }
  throw new CapabilityError("unavailable", "no canvas is available to decode images");
}

/**
 * Decode a blob into something drawable, with its intrinsic size.
 *
 * @param {Blob} blob  The image bytes.
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{source: *, width: number, height: number}>}
 * @throws {CapabilityError} decode_failed — when the bytes are not an image.
 */
async function decode(blob, deps) {
  const injected = deps && /** @type {any} */ (deps).decodeImage;
  if (typeof injected === "function") return injected(blob);
  const g = /** @type {any} */ (globalThis);
  if (typeof g.createImageBitmap === "function") {
    try {
      const bitmap = await g.createImageBitmap(blob);
      return { source: bitmap, width: bitmap.width, height: bitmap.height };
    } catch (err) {
      throw new CapabilityError("decode_failed", err && err.message);
    }
  }
  throw new CapabilityError("decode_failed", "no image decoder is available");
}

/**
 * Encode a canvas at one quality, as a blob.
 *
 * @param {*} canvas   The canvas to read.
 * @param {string} type     The MIME type to encode to.
 * @param {number} quality  The encoder quality, 0 to 1.
 * @returns {Promise<Blob>} The encoded bytes.
 * @throws {CapabilityError} encode_failed — when the browser refuses the type.
 */
async function encode(canvas, type, quality) {
  if (typeof canvas.convertToBlob === "function") {
    return canvas.convertToBlob({ type, quality });
  }
  if (typeof canvas.toBlob === "function") {
    const blob = await new Promise((resolve) => canvas.toBlob(resolve, type, quality));
    if (blob) return blob;
    throw new CapabilityError("encode_failed", `cannot encode to ${type}`);
  }
  throw new CapabilityError("encode_failed", "the canvas cannot be encoded");
}

/**
 * Pick the output type, honouring `"auto"`.
 *
 * `"auto"` is one field rather than the SDK's two capabilities
 * (`bestSupportedType` / `supportsImageType`): the caller wants a small file, not
 * a survey of what the browser encodes.
 *
 * @param {string} requested  `"auto"`, or an explicit MIME type.
 * @param {*} canvas          A canvas to probe with.
 * @returns {Promise<string>} The type to encode to.
 */
async function chooseType(requested, canvas) {
  if (requested && requested !== "auto") return requested;
  try {
    const probe = await encode(canvas, PREFERRED_TYPE, 0.5);
    if (probe && probe.type === PREFERRED_TYPE) return PREFERRED_TYPE;
  } catch {
    // fall through — a browser that cannot encode webp is not an error
  }
  return FALLBACK_TYPE;
}

/**
 * Draw a decoded image onto a canvas at a target size.
 *
 * @param {{source: *, width: number, height: number}} image  The decoded image.
 * @param {number} width   Target width.
 * @param {number} height  Target height.
 * @param {import("./index.js").NativeDeps} deps
 * @returns {*} The canvas holding the drawn image.
 * @throws {CapabilityError} decode_failed — when the canvas has no 2D context.
 */
function draw(image, width, height, deps) {
  const canvas = canvasFactory(deps)(Math.max(1, Math.round(width)), Math.max(1, Math.round(height)));
  const ctx = canvas.getContext("2d");
  if (!ctx) throw new CapabilityError("decode_failed", "no 2D context on the canvas");
  ctx.drawImage(image.source, 0, 0, canvas.width, canvas.height);
  return canvas;
}

/**
 * Scale a size down to fit a maximum edge, keeping the aspect ratio.
 *
 * @param {number} width      Current width.
 * @param {number} height     Current height.
 * @param {?number} maxWidth  The cap, or null for none.
 * @param {?number} maxHeight The cap, or null for none.
 * @returns {{width: number, height: number}} The fitted size, never upscaled.
 */
function fit(width, height, maxWidth, maxHeight) {
  let scale = 1;
  if (maxWidth && width > maxWidth) scale = Math.min(scale, maxWidth / width);
  if (maxHeight && height > maxHeight) scale = Math.min(scale, maxHeight / height);
  return { width: width * scale, height: height * scale };
}

/**
 * Shrink an image to fit a byte budget, by binary search on encoder quality.
 *
 * The search is the whole point. Encoded size is not linear in quality, so a
 * fixed ladder either misses the budget or gives away quality that would have
 * fit. Each step halves the remaining quality range: too big, search lower; fits,
 * keep it and search higher for something better.
 *
 * @param {{source: *, max_kb?: number, max_width?: number, max_height?: number,
 *          type?: string, min_quality?: number, max_quality?: number,
 *          steps?: number}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{ref: string, width: number, height: number, size_kb: number,
 *          quality: number, attempts: number, within_budget: boolean, mime_type: string}>}
 * @throws {CapabilityError} not_found — the handle is unknown or was evicted.
 */
export async function imagingCompress(args, deps) {
  const blob = resolveSource(args.source);
  if (!blob) throw new CapabilityError("not_found", "the image source could not be resolved");

  const image = await decode(blob, deps);
  const target = fit(image.width, image.height, args.max_width, args.max_height);
  const canvas = draw(image, target.width, target.height, deps);
  const type = await chooseType(args.type, canvas);

  const budget = typeof args.max_kb === "number" ? args.max_kb * 1024 : Infinity;
  const steps = args.steps ?? DEFAULT_COMPRESS_STEPS;
  let low = args.min_quality ?? DEFAULT_MIN_QUALITY;
  let high = args.max_quality ?? DEFAULT_MAX_QUALITY;

  let attempts = 0;
  /** @type {?{blob: Blob, quality: number}} */
  let best = null;
  /** @type {?{blob: Blob, quality: number}} */
  let smallest = null;

  for (let step = 0; step < steps && low <= high; step += 1) {
    const quality = (low + high) / 2;
    const encoded = await encode(canvas, type, quality);
    attempts += 1;
    if (!smallest || encoded.size < smallest.blob.size) {
      smallest = { blob: encoded, quality };
    }
    if (encoded.size <= budget) {
      best = { blob: encoded, quality };
      low = quality + 0.01;
    } else {
      high = quality - 0.01;
    }
  }

  const chosen = best ?? smallest;
  if (!chosen) throw new CapabilityError("encode_failed", "no encode attempt succeeded");

  return {
    ref: putBlob(chosen.blob),
    width: Math.round(canvas.width),
    height: Math.round(canvas.height),
    size_kb: Math.round((chosen.blob.size / 1024) * 10) / 10,
    quality: Math.round(chosen.quality * 100) / 100,
    attempts,
    within_budget: chosen.blob.size <= budget,
    mime_type: type,
  };
}

/**
 * Render one image at several sizes, for previews.
 *
 * @param {{source: *, sizes?: number[], type?: string, quality?: number}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{thumbnails: Object[]}>} One entry per requested size, in the
 *          order asked. An empty `sizes` answers an empty list — asking for no
 *          thumbnails is a valid thing to do, not an error.
 * @throws {CapabilityError} not_found — the handle is unknown or was evicted.
 */
export async function imagingThumbnails(args, deps) {
  const blob = resolveSource(args.source);
  if (!blob) throw new CapabilityError("not_found", "the image source could not be resolved");
  const sizes = Array.isArray(args.sizes) ? args.sizes : [];
  if (sizes.length === 0) return { thumbnails: [] };

  const image = await decode(blob, deps);
  const quality = typeof args.quality === "number" ? args.quality : DEFAULT_MAX_QUALITY;
  const thumbnails = [];

  for (const size of sizes) {
    const target = fit(image.width, image.height, size, size);
    const canvas = draw(image, target.width, target.height, deps);
    const type = await chooseType(args.type, canvas);
    const encoded = await encode(canvas, type, quality);
    thumbnails.push({
      ref: putBlob(encoded),
      size,
      width: Math.round(canvas.width),
      height: Math.round(canvas.height),
      size_kb: Math.round((encoded.size / 1024) * 10) / 10,
      mime_type: type,
    });
  }
  return { thumbnails };
}

/**
 * Resize, rotate, crop and flip in a single pass.
 *
 * One capability rather than four, because four means four decodes and four
 * encodes of the same image to do what one canvas does in one.
 *
 * @param {{source: *, width?: number, height?: number, rotate?: number,
 *          flip_horizontal?: boolean, flip_vertical?: boolean,
 *          crop?: {x: number, y: number, width: number, height: number},
 *          type?: string, quality?: number}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<Object>} The processed image's handle and shape.
 * @throws {CapabilityError} not_found — the handle is unknown or was evicted.
 */
export async function imagingTransform(args, deps) {
  const blob = resolveSource(args.source);
  if (!blob) throw new CapabilityError("not_found", "the image source could not be resolved");
  const image = await decode(blob, deps);

  const crop = args.crop;
  const cropped = crop
    ? {
        source: image.source,
        width: Math.min(crop.width, image.width - crop.x),
        height: Math.min(crop.height, image.height - crop.y),
        x: crop.x,
        y: crop.y,
      }
    : { source: image.source, width: image.width, height: image.height, x: 0, y: 0 };

  const target = fit(cropped.width, cropped.height, args.width, args.height);
  const rotate = ((args.rotate || 0) % 360 + 360) % 360;
  const swapped = rotate === 90 || rotate === 270;
  const canvasWidth = swapped ? target.height : target.width;
  const canvasHeight = swapped ? target.width : target.height;

  const canvas = canvasFactory(deps)(
    Math.max(1, Math.round(canvasWidth)),
    Math.max(1, Math.round(canvasHeight)),
  );
  const ctx = canvas.getContext("2d");
  if (!ctx) throw new CapabilityError("decode_failed", "no 2D context on the canvas");

  ctx.save();
  ctx.translate(canvas.width / 2, canvas.height / 2);
  if (rotate) ctx.rotate((rotate * Math.PI) / 180);
  ctx.scale(args.flip_horizontal ? -1 : 1, args.flip_vertical ? -1 : 1);
  ctx.drawImage(
    cropped.source,
    cropped.x,
    cropped.y,
    cropped.width,
    cropped.height,
    -target.width / 2,
    -target.height / 2,
    target.width,
    target.height,
  );
  ctx.restore();

  const type = await chooseType(args.type, canvas);
  const quality = typeof args.quality === "number" ? args.quality : DEFAULT_MAX_QUALITY;
  const encoded = await encode(canvas, type, quality);

  return {
    ref: putBlob(encoded),
    width: Math.round(canvas.width),
    height: Math.round(canvas.height),
    size_kb: Math.round((encoded.size / 1024) * 10) / 10,
    mime_type: type,
  };
}

/**
 * Report an image's type, size and dimensions.
 *
 * Cheap by design: the byte count and MIME come from the blob itself, and only
 * the dimensions need a decode. An app deciding whether an image is worth
 * compressing at all asks this first.
 *
 * @param {{source: *}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{mime_type: string, width: number, height: number, size_kb: number}>}
 * @throws {CapabilityError} not_found — the handle is unknown or was evicted.
 */
export async function imagingInfo(args, deps) {
  const blob = resolveSource(args.source);
  if (!blob) throw new CapabilityError("not_found", "the image source could not be resolved");
  const described = typeof args.source === "string" ? describeBlob(args.source) : null;
  const image = await decode(blob, deps);
  return {
    mime_type: (described && described.mime) || blob.type || "application/octet-stream",
    width: image.width,
    height: image.height,
    size_kb: Math.round((blob.size / 1024) * 10) / 10,
  };
}

/**
 * Read an image's bytes back into Python.
 *
 * The escape hatch, not the path. Calling this moves the whole image across the
 * bridge, which is what handles exist to avoid — an app that only needs to upload
 * should hand the handle to `http.upload` instead.
 *
 * @param {{source: *}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{data_base64: string, mime_type: string, size_kb: number}>}
 * @throws {CapabilityError} not_found — the handle is unknown or was evicted.
 */
export async function imagingRead(args, deps) {
  const blob = resolveSource(args.source);
  if (!blob) throw new CapabilityError("not_found", "the image source could not be resolved");
  return {
    data_base64: await toBase64(blob),
    mime_type: blob.type || "application/octet-stream",
    size_kb: Math.round((blob.size / 1024) * 10) / 10,
  };
}

/**
 * Release a handle, or every handle.
 *
 * @param {{source?: *, all?: boolean}} args
 * @param {import("./index.js").NativeDeps} _deps
 * @returns {Promise<{released: number}>} How many handles were released.
 */
export async function imagingRelease(args, _deps) {
  if (args && args.all) return { released: clearBlobs() };
  if (args && typeof args.source === "string") {
    return { released: dropBlob(args.source) ? 1 : 0 };
  }
  return { released: 0 };
}
